### PR TITLE
EN-65713: move release tagging

### PR DESCRIPTION
**Pull request by sarah-salinas**
This commit moves the release tagging after a successful docker build. In order to do this, it separates the docker build from the publishing of the images into separate Jenkinsfile steps.

### CODE REVIEW

**Comment by sarah-salinas** [comment]
## Testing

### Simulated Main Build

https://jenkins-build.socrata.com/job/soda-fountain-branch/18/console

### Release Dry Run

TODO

### Simulated Release Build

TODO

**Review by rjmac — APPROVED** [review]

File: Jenkinsfile [hunk 1]
```diff
@@ -71,52 +78,63 @@ pipeline {
...
+        success {
+          script {
+            if (params.RELEASE_BUILD && !params.RELEASE_DRY_RUN){
+              env.GIT_TAG = releaseTag.getFormattedTag(params.RELEASE_NAME)
+              if (releaseTag.doesReleaseTagExist(params.RELEASE_NAME)) {
+                echo "REBUILD: Tag ${env.GIT_TAG} already exists"
+                return
+              }
+              if (params.RELEASE_DRY_RUN) {
+                echo "DRY RUN: Would have created ${env.GIT_TAG} and pushed it to the repo"
```
Comment (ds2643) [inline Jenkinsfile:109]: Is this statement reachable? The test conditions seem mutually exclusive, unless I'm misreading the nesting.

File: Jenkinsfile [hunk 2]
```diff
@@ -71,52 +78,63 @@ pipeline {
...
+          // Passing [] to overrideEnvironments will not override the environments, so the docker image will be published to all environments
+          Set environments = (params.RELEASE_BUILD) ? [] : ['internal']
+          dockerize.publish(sourceImage: env.DOCKER_TAG, overrideEnvironments: environments)
         }
       }
       post {
         success {
           script {
-            if (params.RELEASE_BUILD && !params.RELEASE_DRY_RUN) {
+            if (params.RELEASE_BUILD) {
```
Comment (ds2643) [inline Jenkinsfile:137]: Are RELEASE_BUILD and RELEASE_DRY_RUN now mutually exclusive? How are we certain dry runs won't write builds to RMS?

**Review by ds2643 — COMMENTED** [review]
See questions.

File: Jenkinsfile (hunk 1 above)
Comment (sarah-salinas) [reply Jenkinsfile:109]: I forgot to edit line 102.  I will fix the logic.

File: Jenkinsfile (hunk 2 above)
Comment (sarah-salinas) [reply Jenkinsfile:137]: See line 123 -- if it's a RELEASE_BUILD && RELEASE_DRY_RUN, it does not execute the Publish stage.

File: Jenkinsfile (hunk 1 above)
Comment (sarah-salinas) [reply Jenkinsfile:109]: Updated

File: Jenkinsfile (hunk 2 above)
Comment (sarah-salinas) [reply Jenkinsfile:137]: RELEASE_DRY_RUN means nothing if RELEASE_BUILD is not true.

File: Jenkinsfile [hunk 3]
```diff
@@ -36,19 +37,25 @@ pipeline {
     WEBHOOK_ID = 'WEBHOOK_IQ'
   }
   stages {
-    stage('Release Tag') {
+    stage('Publish Library') {
```
Comment (sarah-salinas) [inline Jenkinsfile:40]: I made a ticket to automate the publishing of the library when there's a change to the verison.sbt file
https://socrata.atlassian.net/browse/EN-65991

**Comment by sarah-salinas** [comment]
## Testing

### Simulated Main Build

https://jenkins-build.socrata.com/job/soda-fountain-branch/21/console - 🟢 

Stages
- Build
- Docker Build
- Publish Image: internal only
- Deploy: staging

Description: `soda-fountain:STAGING_21_e3be5a93`

### Release Dry Run

https://jenkins-build.socrata.com/job/soda-fountain-branch/22/console - 🟢 

Stages
- Build
- Docker Build

Description: `soda-fountain:H23 - DRY RUN`

### Simulated Release Build

https://jenkins-build.socrata.com/job/soda-fountain-branch/23/console - 🟢 

Stages
- Build
- Docker Build , Post: release tag
- Publish Image: all environments, Post: rms create build

Description: `soda-fountain:H23`

**Review by rjmac — APPROVED** [review]

File: Jenkinsfile [hunk 4]
```diff
@@ -18,13 +18,16 @@ def releaseTag = new com.socrata.ReleaseTag(steps, service)
 pipeline {
   options {
     ansiColor('xterm')
+    buildDiscarder(logRotator(numToKeepStr: '50'))
+    disableConcurrentBuilds(abortPrevious: true)
+    timeout(time: 20, unit: 'MINUTES')
   }
   parameters {
+    string(name: 'AGENT', defaultValue: 'build-worker', description: 'Which build agent to use?')
```
Comment (ds2643) [inline Jenkinsfile:26]: Again, not a question, but this comment is very minor and non-functional.

File: Jenkinsfile (hunk 1 above)
Comment (ds2643) [reply Jenkinsfile:115]: Thanks

File: Jenkinsfile [hunk 5]
```diff
@@ -141,17 +172,23 @@ pipeline {
...
         }
       }
     }
   }
   post {
     failure {
       script {
-        if (!isPr) {
-          teamsMessage(message: "Build [${currentBuild.fullDisplayName}](${env.BUILD_URL}) has failed in stage ${lastStage}", webhookCredentialID: WEBHOOK_ID)
+        if (env.JOB_NAME == "${service}/main") {
```
Comment (ds2643) [inline Jenkinsfile:187]: I suggest binding `(env.JOB_NAME == "${service}/main")` to a descriptive variable describing the meaning of the condition (i.e., `isRoutinePullRequest`, if that's what this condition means).

**Review by ds2643 — APPROVED** [review]
These changes reflect those made in other repositories. This work helps improve logistics around build failure by avoiding unnecessary complications.

**Pull request merged**